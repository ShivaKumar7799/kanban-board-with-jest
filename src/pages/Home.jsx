import React, { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import { signInUser } from "../redux/features/userSlice";
import { useNavigate, Link } from "react-router-dom";
import { useFormik } from "formik";
import apiService from "../Services/Api Services/ApiServices";

function Home() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [showPassword, setShowPassword] = useState(false);

  const [users, setUsers] = useState([]);
  const [invalidUser, setInvalidUser] = useState(false);

  useEffect(() => {
    apiService.get(`/signUp`).then((res) => setUsers(res.data)).catch(err => console.log("signup api error"));
  }, []);

  const loginValidation = (values) => {
    setInvalidUser(false);
    const errors = {};

    if (values.userNameOrEmail === "") {
      errors.userNameOrEmail = "UserName/email is Mandatory";
    }

    if (values.password === "") {
      errors.password = "Password is Required";
    } else if (values.password.length < 8) {
      errors.password = "Password should contain minimun 8 characters";
    } else if (values.password.length > 20) {
      errors.password = "password shouldn't contain more than 20 characters";
    }

    return errors;
  };

  const loginFormik = useFormik({
    initialValues: {
      userNameOrEmail: "",
      password: "",
    },
    validate: loginValidation,
    onSubmit: (values) => {
      const validUser = users.filter(
        (user) =>
          (user.userName === values.userNameOrEmail ||
            user.email === values.userNameOrEmail) &&
          user.password === values.password
      );
      if (validUser.length > 0) {
        navigate("/dashBoard");
        dispatch(signInUser(validUser[0]));
        localStorage.setItem("user", JSON.stringify(validUser[0]));
      } else {
        setInvalidUser(true);
      }
    },
  });
  return (
    <div className="d-flex justify-content-center min-vh-100 bg-dark align-items-center">
      <div
        className="w-50 border border-primary rounded py-3 px-5 bg-white h-75"
        style={{ minHeight: "375px" }}
      >
        <h1 className="text-center text-dark"> Kanban Board</h1>

        <form onSubmit={loginFormik.handleSubmit}>
          <div className="d-flex flex-column">
            <label htmlFor="userNameOrEmail">Username/email</label>
            <input
              type="text"
              name="userNameOrEmail"
              id="userNameOrEmail"
              className="form-control"
              placeholder="Please enter your username or email here"
              value={loginFormik.userNameOrEmail}
              onChange={loginFormik.handleChange}
              onBlur={loginFormik.handleBlur}
            />
            {loginFormik.errors.userNameOrEmail &&
              loginFormik.touched.userNameOrEmail && (
                <span className="text-danger py-2">
                  {loginFormik.errors.userNameOrEmail}
                </span>
              )}
          </div>

          <div className="d-flex flex-column">
            <label htmlFor="password">password*</label>
            <input
              type={`${showPassword ? "text" : "password"}`}
              value={loginFormik.password}
              className="form-control"
              name="password"
              id="password"
              placeholder="Enter your password"
              onChange={loginFormik.handleChange}
              onBlur={loginFormik.handleBlur}
            />
            <span
              className="text-primary cursor-pointer"
              onClick={() => setShowPassword(!showPassword)}
            >
              {showPassword ? "hide password" : "show password"}{" "}
            </span>
            {loginFormik.errors.password && loginFormik.touched.password && (
              <span className="text-danger py-2">
                {loginFormik.errors.password}
              </span>
            )}
          </div>

          <div>
            {invalidUser && (
              <div className="text-danger pt-2">
                Invalid Username or Password
              </div>
            )}
          </div>

          <div className="d-flex justify-content-center py-3">
            <input
              className="btn btn-success btn-lg"
              type="submit"
              value="Login"
            />
          </div>

          <div>
            <span>
              Don't have account? <Link to="/signUp">Register Here</Link>
            </span>
          </div>
        </form>
      </div>
    </div>
  );
}

export default Home;
