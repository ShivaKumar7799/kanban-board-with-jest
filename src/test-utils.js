
// test-utils.js
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import {reduxStore} from './redux/store/store'; // Import your Redux store

const render = (ui, { store = reduxStore, ...renderOptions } = {}) => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

// Re-export everything from @testing-library/react
export * from '@testing-library/react';

// Override render method
export { render };