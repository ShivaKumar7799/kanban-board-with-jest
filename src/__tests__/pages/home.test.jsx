import userEvent from "@testing-library/user-event"
import { render, screen } from "../../test-utils"
import { act } from 'react-dom/test-utils';
import Home from "../../pages/Home"
import React from "react"
import { MemoryRouter } from "react-router-dom"

const setupHomeComponent = () => {
  render(
    <MemoryRouter>
      <Home />
    </MemoryRouter>
  );
};

describe('renders home componenet correctly', () => {

  test('loader component snapshot', () => {
    const { container } = render(<MemoryRouter>
      <Home />
    </MemoryRouter>);
    expect(container).toMatchSnapshot();
  });

  test("display Kanban Board Title correctly", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const titleElement = screen.getByRole("heading", {
      level: 1
    });
    expect(titleElement).toBeInTheDocument();
  })

  test("render Username/email label correctly", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const labelElement = screen.getByLabelText(/username\/email/i);
    expect(labelElement).toBeInTheDocument();
  })

  test("render userNameOrEmail input field correctly", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const userNameOrEmailInputElement = screen.getByRole("textbox", {
      name: "Username/email"
    })
    expect(userNameOrEmailInputElement).toBeInTheDocument();
    expect(userNameOrEmailInputElement).toBeEnabled();
    expect(userNameOrEmailInputElement).toHaveAttribute("type", "text")
  })

  test("render password label correctly", () => {
   // eslint-disable-next-line testing-library/no-unnecessary-act
   act(() => {
    setupHomeComponent()
  })
    const passwordElement = screen.getByText('password*');
    expect(passwordElement).toBeInTheDocument();
  })

  test("renders password input field correctly", () => {
   // eslint-disable-next-line testing-library/no-unnecessary-act
   act(() => {
    setupHomeComponent()
  })
    const passwordInputElement = screen.getByPlaceholderText("Enter your password")
    expect(passwordInputElement).toBeInTheDocument();
    expect(passwordInputElement).toBeEnabled();
    expect(passwordInputElement).toHaveAttribute('type', 'password');
  } )

  test("login button renders correctly", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const buttonElement = screen.getByRole("button", {
      name: "Login"
    })
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).not.toBeDisabled();
  })

  test("changing username/email and password input field value", async() => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const userNameOrEmailInputElement = screen.getByRole("textbox", {
      name: "Username/email"
    })
    const passwordInputElement = screen.getByPlaceholderText("Enter your password")
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.type(userNameOrEmailInputElement, "shiva kumar")
    })
    expect(userNameOrEmailInputElement).toHaveValue("shiva kumar")

     // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.type(passwordInputElement, "Shiva@123")
    })
    expect(passwordInputElement).toHaveValue("Shiva@123")
  })

  test("clicking on username/email label should focus input field", async() => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const labelElement = screen.getByLabelText(/username\/email/i);
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(labelElement)
    })
    const inputElement = screen.getByRole("textbox",{
      name: "Username/email"
    })
    await expect(inputElement).toHaveFocus();
  })

  test("clicking on password label should focus password input field", async() => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      setupHomeComponent()
    })
    const passwordLabelElement = screen.getByText('password*');
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(passwordLabelElement)
    })
    const passwordInputElement = screen.getByPlaceholderText("Enter your password")
    await expect(passwordInputElement).toHaveFocus();
  })

  test("showinig error when user focus on input field and focussing out without entering the value", async() => {
    act(() => {
      setupHomeComponent();
    })

    const userNameOrEmailInputElement = screen.getByRole("textbox", {
      name: "Username/email"
    })
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(userNameOrEmailInputElement);
    })
    expect(userNameOrEmailInputElement).toHaveFocus();

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(document.body)
    })

     expect(userNameOrEmailInputElement).not.toHaveFocus();
     const userNameOrEmailErrorElement = screen.getByText("UserName/email is Mandatory");
     expect(userNameOrEmailErrorElement).toBeInTheDocument();
  })

  test("entering password length less than 8 charcters should show error", async() => {
    act(() => {
      setupHomeComponent();
    })
    const passwordInputElement = screen.getByPlaceholderText("Enter your password");
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.type(passwordInputElement,"Shiva")
    })
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(document.body)
    })
    const passwordErrorElement = screen.getByText("Password should contain minimun 8 characters")
    await expect(passwordErrorElement).toBeInTheDocument();
  })

  test("entering password more than 20 charcters should show error", async() => {
    act(() => {
      setupHomeComponent();
    })
    const passwordInputElement = screen.getByPlaceholderText("Enter your password");
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.type( passwordInputElement, "abcdabcdabcdabcdabcda")
    })
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(document.body)
    });
    const passwordErrorElement = screen.getByText(/contain more than 20 characters/i)
    await expect(passwordErrorElement).toBeInTheDocument();
  })

  test("clicking on showPassword should show hide password text", async() => {
    act(() => {
      setupHomeComponent();
    })
    const showPasswordElement = screen.getByText("show password");
    expect(showPasswordElement).toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(() => {
      userEvent.click(showPasswordElement);
    })

    const hidePasswordElement = screen.getByText("hide password")
    expect(hidePasswordElement).toBeInTheDocument();
  })

  test("rendering Don't have account? text correctly", () => {
    act(() => {
      setupHomeComponent();
    } )
    const noAccountExistsElement = screen.getByText("Don't have account?");
    expect(noAccountExistsElement).toBeInTheDocument();
  })

})
