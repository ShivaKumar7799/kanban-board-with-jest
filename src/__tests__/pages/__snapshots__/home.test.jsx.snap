// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders home componenet correctly loader component snapshot 1`] = `
<div>
  <div
    class="d-flex justify-content-center min-vh-100 bg-dark align-items-center"
  >
    <div
      class="w-50 border border-primary rounded py-3 px-5 bg-white h-75"
      style="min-height: 375px;"
    >
      <h1
        class="text-center text-dark"
      >
         Kanban Board
      </h1>
      <form>
        <div
          class="d-flex flex-column"
        >
          <label
            for="userNameOrEmail"
          >
            Username/email
          </label>
          <input
            class="form-control"
            id="userNameOrEmail"
            name="userNameOrEmail"
            placeholder="Please enter your username or email here"
            type="text"
            value=""
          />
        </div>
        <div
          class="d-flex flex-column"
        >
          <label
            for="password"
          >
            password*
          </label>
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="Enter your password"
            type="password"
            value=""
          />
          <span
            class="text-primary cursor-pointer"
          >
            show password
             
          </span>
        </div>
        <div />
        <div
          class="d-flex justify-content-center py-3"
        >
          <input
            class="btn btn-success btn-lg"
            type="submit"
            value="Login"
          />
        </div>
        <div>
          <span>
            Don't have account? 
            <a
              href="/signUp"
            >
              Register Here
            </a>
          </span>
        </div>
      </form>
    </div>
  </div>
</div>
`;
